File: forge/src/main/java/org/valkyrienskies/Tournament/forge/TournamentModForge.java
package org.valkyrienskies.tournament.forge;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.world.item.CreativeModeTab;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.client.ConfigGuiHandler;
import net.minecraftforge.client.event.EntityRenderersEvent;
import net.minecraftforge.client.event.ModelRegistryEvent;
import net.minecraftforge.eventbus.api.IEventBus;
import net.minecraftforge.fml.ModLoadingContext;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import org.jetbrains.annotations.NotNull;
import org.valkyrienskies.core.impl.config.VSConfigClass;
import org.valkyrienskies.tournament.TournamentBlocks;
import org.valkyrienskies.tournament.TournamentConfig;
import org.valkyrienskies.tournament.TournamentItems;
import org.valkyrienskies.tournament.TournamentMod;
import org.valkyrienskies.mod.compat.clothconfig.VSClothConfig;

@Mod(TournamentMod.MOD_ID)
public class
TournamentModForge {
    boolean happendClientSetup = false;
    static IEventBus MOD_BUS;

    public TournamentModForge() {
        // Submit our event bus to let architectury register our content on the right time
        MOD_BUS = FMLJavaModLoadingContext.get().getModEventBus();
        MOD_BUS.addListener(this::clientSetup);

        ModLoadingContext.get().registerExtensionPoint(ConfigGuiHandler.ConfigGuiFactory.class,
                () -> new ConfigGuiHandler.ConfigGuiFactory((Minecraft client, Screen parent) ->
                        VSClothConfig.createConfigScreenFor(parent,
                                VSConfigClass.Companion.getRegisteredConfig(TournamentConfig.class)))
        );

        MOD_BUS.addListener(this::onModelRegistry);
        MOD_BUS.addListener(this::clientSetup);
        MOD_BUS.addListener(this::entityRenderers);

        TournamentItems.TAB = new CreativeModeTab("vs_tournament.main_tab") {
            @Override
            public @NotNull ItemStack makeIcon() {
                return new ItemStack(TournamentBlocks.INSTANCE.getSHIP_ASSEMBLER().get());
            }
        };

        TournamentMod.init();
    }

    void clientSetup(final FMLClientSetupEvent event) {
        if (happendClientSetup) return;
        happendClientSetup = true;

        TournamentMod.initClient();
    }

    void entityRenderers(final EntityRenderersEvent.RegisterRenderers event) {
    }

    void onModelRegistry(final ModelRegistryEvent event) {
    }
}

File: forge/src/main/kotlin/org/valkyrienskies/tournament/forge/TournamentModForge.kt
package org.valkyrienskies.tournament.forge

import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider
import net.minecraft.core.registries.Registries
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraftforge.client.event.EntityRenderersEvent.RegisterRenderers
import net.minecraftforge.client.event.ModelEvent
import net.minecraftforge.event.TickEvent.ServerTickEvent
import net.minecraftforge.eventbus.api.IEventBus
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent
import org.valkyrienskies.tournament.TickScheduler
import org.valkyrienskies.tournament.TournamentMod
import org.valkyrienskies.tournament.TournamentMod.init
import org.valkyrienskies.tournament.TournamentMod.initClient
import org.valkyrienskies.tournament.TournamentMod.initClientRenderers
import org.valkyrienskies.tournament.TournamentModels
import thedarkcolour.kotlinforforge.forge.FORGE_BUS
import thedarkcolour.kotlinforforge.forge.MOD_BUS

@Mod(TournamentMod.MOD_ID)
class TournamentModForge {
    private var happendClientSetup = false

    init {
        FORGE_BUS.addListener { event: ServerTickEvent ->
            TickScheduler.tickServer(event.server)
        }

        // Submit our event bus to let architectury register our content on the right time
        MOD_BUS.addListener { event: FMLClientSetupEvent? ->
            clientSetup(
                event
            )
        }
        FORGE_BUS.addListener { event: ModelEvent? ->
            onModelRegistry(
                event
            )
        }
        MOD_BUS.addListener { event: RegisterRenderers ->
            entityRenderers(
                event
            )
        }
        init()
    }

    private fun clientSetup(event: FMLClientSetupEvent?) {
        if (happendClientSetup) {
            return
        }
        happendClientSetup = true
        initClient()
    }

    private fun entityRenderers(event: RegisterRenderers) {
        initClientRenderers(
            object : TournamentMod.ClientRenderers {
                override fun <T : BlockEntity> registerBlockEntityRenderer(
                    t: BlockEntityType<T>,
                    r: BlockEntityRendererProvider<T>
                ) = event.registerBlockEntityRenderer(t, r)
            }
        )
    }

    private fun onModelRegistry(event: ModelEvent?) {
        println("[Tournament] Registering models")
        TournamentModels.MODELS.forEach { rl ->
            println("[Tournament] Registering model $rl")
            // ForgeModelBakery.addSpecialModel(rl)
        }
    }

    companion object {
        fun getModBus(): IEventBus = MOD_BUS
    }
}